File: algo/src/main/java/com/hartwig/actin/algo/evaluation/treatment/HasHadPDFollowingTreatmentWithCategory.java
package com.hartwig.actin.algo.evaluation.treatment;

import static com.hartwig.actin.algo.evaluation.treatment.ProgressiveDiseaseFunctions.treatmentResultedInPDOption;

import java.util.Optional;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.evaluation.EvaluationFactory;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;
import com.hartwig.actin.clinical.datamodel.PriorTumorTreatment;
import com.hartwig.actin.clinical.datamodel.TreatmentCategory;

import org.jetbrains.annotations.NotNull;

public class HasHadPDFollowingTreatmentWithCategory implements EvaluationFunction {

    @NotNull
    private final TreatmentCategory category;

    HasHadPDFollowingTreatmentWithCategory(@NotNull final TreatmentCategory category) {
        this.category = category;
    }

    @NotNull
    @Override
    public Evaluation evaluate(@NotNull PatientRecord record) {
        boolean hasHadPDFollowingTreatmentWithCategory = false;
        boolean hasHadTreatmentWithUnclearPDStatus = false;
        boolean hasHadTrial = false;

        for (PriorTumorTreatment treatment : record.clinical().priorTumorTreatments()) {
            if (treatment.categories().contains(category)) {
                Optional<Boolean> treatmentResultedInPDOption = treatmentResultedInPDOption(treatment);
                if (treatmentResultedInPDOption.orElse(false)) {
                    hasHadPDFollowingTreatmentWithCategory = true;
                } else if (treatmentResultedInPDOption.isEmpty()) {
                    hasHadTreatmentWithUnclearPDStatus = true;
                }
            }

            if (treatment.categories().contains(TreatmentCategory.TRIAL)) {
                hasHadTrial = true;
            }
        }

        if (hasHadPDFollowingTreatmentWithCategory) {
            return EvaluationFactory.pass("Patient has had progressive disease following treatment with category " + category.display(),
                    category.display() + " treatment with PD");
        } else if (hasHadTreatmentWithUnclearPDStatus) {
            return EvaluationFactory.undetermined(
                    "Patient has had treatment with category " + category.display() + " but unclear PD status",
                    "Had " + category.display() + " treatment with unclear PD status");
        } else if (hasHadTrial) {
            return EvaluationFactory.undetermined("Patient has had trial with unclear treatment category", "Trial treatment");
        } else {
            return EvaluationFactory.fail("Patient has no progressive disease following treatment with category " + category.display(),
                    "No " + category.display() + " treatment with PD");
        }

    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/treatment/HasHadPDFollowingTreatmentWithCategory.kt
package com.hartwig.actin.algo.evaluation.treatment

import com.hartwig.actin.algo.evaluation.EvaluationFactory.fail
import com.hartwig.actin.algo.evaluation.EvaluationFactory.pass
import com.hartwig.actin.algo.evaluation.EvaluationFactory.undetermined
import com.hartwig.actin.algo.evaluation.EvaluationFunction
import com.hartwig.actin.clinical.interpretation.ProgressiveDiseaseFunctions
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.algo.Evaluation
import com.hartwig.actin.datamodel.clinical.treatment.TreatmentCategory

class HasHadPDFollowingTreatmentWithCategory(private val category: TreatmentCategory) : EvaluationFunction {

    override fun evaluate(record: PatientRecord): Evaluation {
        val history = record.oncologicalHistory
        val treatmentSummary = TreatmentSummaryForCategory.createForTreatmentHistory(
            history,
            category,
            { entry -> ProgressiveDiseaseFunctions.treatmentResultedInPD(entry, history) },
            { true },
            { entry -> ProgressiveDiseaseFunctions.treatmentResultedInPD(entry, history) != false }
        )

        return if (treatmentSummary.hasSpecificMatch()) {
            pass(category.display() + " treatment in provided treatments with PD")
        } else if (treatmentSummary.hasApproximateMatch()) {
            undetermined(category.display() + " treatment in provided treatments but PD status is undetermined")
        } else if (treatmentSummary.hasPossibleTrialMatch()) {
            undetermined("Undetermined if treatment from previous trial included $category")
        } else {
            fail("No " + category.display() + " treatment with PD in provided treatments")
        }
    }
}